File: infrastructure/typescript/resources/utils/websocket-broadcaster.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { MetricUnits } from '@aws-lambda-powertools/metrics';

// Helper class to send websocket messages to ALL connected users.
export class WebsocketBroadcaster {

    constructor(private AWS: any,
        private metrics: any,
        private dynamoDbClient: any,
        private logger: any,
        private connectionsTableName: string) { }

    private _apiGatewayEndpoint!: string;
    private _apigwManagementApi: any;

    async broadcast(payload: any, apiGatewayEndpoint: string) {
        try {

            this.logger.debug('Retrieving active connections...');
            let connectionData = await this.dynamoDbClient.scan({ TableName: this.connectionsTableName, ProjectionExpression: 'connectionId' }).promise();
            this.logger.debug('ConnectionData:', connectionData);
            this.logger.debug(`Cached ApiGatewayEndpoint: ${this._apiGatewayEndpoint}`);
            this.logger.debug(`New ApiGatewayEndpoint: ${apiGatewayEndpoint}`);

            this._apigwManagementApi = new this.AWS.ApiGatewayManagementApi({ apiVersion: '2018-11-29', endpoint: apiGatewayEndpoint });

            await Promise.all(connectionData.Items.map(async (connectionData: any) => {
                this.logger.debug(`Sending message to ${connectionData.connectionId}`);
                await this._apigwManagementApi.postToConnection({ ConnectionId: connectionData.connectionId, Data: JSON.stringify(payload) }).promise()
                    .then(()=> {
                        this.metrics.addMetric('messageDelivered', MetricUnits.Count, 1);
                        this.logger.debug(`Message sent to connection ${connectionData.connectionId}`);
                    })
                    .catch((err: any) => {
                        this.logger.debug(`Error during message delivery: ${JSON.stringify(err)}`);
                        if (err.statusCode === 410) {
                            this.logger.debug(`Found stale connection, deleting ${connectionData.connectionId}`);
                            this.dynamoDbClient.delete({ TableName: this.connectionsTableName, Key: { connectionData } });
                        }
                    });
            }));
            this.logger.debug(`All messages have been broadcasted.`);

        } catch (err: any) {
            this.logger.debug("ERROR:", err);
        }
    }
}

File: infrastructure-net7/src/Shared/WebsocketBroadcaster.cs
using System.Net;
using System.Text;
using Amazon.ApiGatewayManagementApi;
using Amazon.ApiGatewayManagementApi.Model;
using Amazon.DynamoDBv2.DataModel;
using Amazon.Runtime;
using AWS.Lambda.Powertools.Logging;
using AWS.Lambda.Powertools.Metrics;
using AWS.Lambda.Powertools.Tracing;
using Shared.Models;

namespace Shared;

public class WebsocketBroadcaster
{
    private readonly IDynamoDBContext _dbContext;
    public WebsocketBroadcaster(IDynamoDBContext dbContext)
    {
        this._dbContext = dbContext;
    }

    [Logging(LogEvent = true, Service = "websocketMessagingService")]
    public async Task Broadcast(string payload, string apiGatewayEndpoint)
    {
        Logger.LogInformation("[Broadcaster] - Retrieving active connections...");
        List<Connection> connectionData;
        try
        {
            connectionData = await _dbContext.ScanAsync<Connection>(Array.Empty<ScanCondition>()).GetRemainingAsync();
            Logger.LogInformation("Retrieved active connections");
            Logger.LogInformation(connectionData);
            
            Logger.LogInformation("Encoding payload to binary:");
            var messageBinary = UTF8Encoding.UTF8.GetBytes(payload);
            Logger.LogInformation(messageBinary);

            // Broadcast message parallel with concurrency limit to avoid API throttling
            var options = new ParallelOptions { MaxDegreeOfParallelism = 5 };
            try
            {
                await Parallel.ForEachAsync(connectionData, options, async (item, token) =>
                {
                    try
                    {
                        Logger.LogInformation($"Broadcasting connection item: {item.connectionId} - {item.userId}");
                        var apiClient = new AmazonApiGatewayManagementApiClient(new AmazonApiGatewayManagementApiConfig
                        {
                            ServiceURL = apiGatewayEndpoint
                        });
                        //Logger.LogInformation(apiClient);
                        var stream = new MemoryStream(messageBinary);
                        var postConnectionRequest = new PostToConnectionRequest
                        {
                            ConnectionId = item.connectionId,
                            Data = stream
                        };

                        Logger.LogInformation($"Broadcast to connection: {item.connectionId}");
                        await apiClient.PostToConnectionAsync(postConnectionRequest);
                        
                        Metrics.AddMetric("messageDelivered", 1, MetricUnit.Count);
                    }
                    catch (AmazonServiceException e)
                    {
                        // API Gateway returns a status of 410 GONE when the connection is no
                        // longer available. If this happens, we simply delete the identifier
                        // from our DynamoDB table.
                        if (e.StatusCode == HttpStatusCode.Gone)
                        {
                            Logger.LogInformation($"Deleting stale connection: {item.connectionId}");
                            await _dbContext.DeleteAsync(item);
                        }
                        else
                        {
                            Logger.LogError($"Error posting message to {item.connectionId}: {e.Message}");
                            Logger.LogCritical(e.StackTrace);
                        }
                    }
                    catch (Exception e)
                    {
                        Logger.LogInformation("Error while broadcasting messages!");
                        Logger.LogError(e);
                        Logger.LogError(e.StackTrace);
                    }
                });
            }
            catch (Exception e)
            {
                Logger.LogError($"Error retrieving connections: {e.Message}");
                Logger.LogCritical(e.StackTrace);
            }
        
        }
        catch (Exception e)
        {
            Logger.LogInformation("Error while processing messages in parallel!");
            Logger.LogError(e);
            Logger.LogError(e.StackTrace);
        }
    }
}